Deployment Failure Reason: spawn EBUSY

Before filing this issue, please ensure you're using the latest CLI by running `swa --version` and comparing to the latest version on [npm](https://www.npmjs.com/package/@azure/static-web-apps-cli).


**Are you accessing the CLI from the default port `:4280` ?**

- [ ] No, I am using a different port number (`--port`) and accessing the CLI from that port
- [ ] Yes, I am accessing the CLI from port `:4280`

> Make sure you are accessing the URL printed in the console when running `swa start`!

**ℹ️ NOTE: Make sure to enable debug logs when running any `swa` commands using `--verbose=silly`**


**Describe the bug**
Sometimes the deployment via `swa deploy` fails on CI.
```
##[debug]Evaluating: secrets.AZURE_STATIC_WEB_APPS_API_TOKEN_MANGO_PEBBLE_0224C3803
##[debug]Evaluating Index:
##[debug]..Evaluating secrets:
##[debug]..=> Object
##[debug]..Evaluating String:
##[debug]..=> 'AZURE_STATIC_WEB_APPS_API_TOKEN_MANGO_PEBBLE_0224C3803'
##[debug]=> '***'
##[debug]Result: '***'
##[debug]Evaluating condition for step: 'Deploy Web App'
##[debug]Evaluating: success()
##[debug]Evaluating success:
##[debug]=> true
##[debug]Result: true
##[debug]Starting: Deploy Web App
##[debug]Loading inputs
##[debug]Evaluating: format('# Delete old deployment (otherwise we get a spawn EBUSY error when running the workflow again)
##[debug]az staticwebapp environment delete --name jabref-online --environment-name {0} --yes || true
##[debug]
##[debug]yarn swa deploy .output\public --env {1}
##[debug]', matrix.deployment_environment, matrix.deployment_environment)
##[debug]Evaluating format:
##[debug]..Evaluating String:
##[debug]..=> '# Delete old deployment (otherwise we get a spawn EBUSY error when running the workflow again)
##[debug]az staticwebapp environment delete --name jabref-online --environment-name {0} --yes || true
##[debug]
##[debug]yarn swa deploy .output\public --env {1}
##[debug]'
##[debug]..Evaluating Index:
##[debug]....Evaluating matrix:
##[debug]....=> Object
##[debug]....Evaluating String:
##[debug]....=> 'deployment_environment'
##[debug]..=> 'default'
##[debug]..Evaluating Index:
##[debug]....Evaluating matrix:
##[debug]....=> Object
##[debug]....Evaluating String:
##[debug]....=> 'deployment_environment'
##[debug]..=> 'default'
##[debug]=> '# Delete old deployment (otherwise we get a spawn EBUSY error when running the workflow again)
##[debug]az staticwebapp environment delete --name jabref-online --environment-name default --yes || true
##[debug]
##[debug]yarn swa deploy .output\public --env default
##[debug]'
##[debug]Result: '# Delete old deployment (otherwise we get a spawn EBUSY error when running the workflow again)
##[debug]az staticwebapp environment delete --name jabref-online --environment-name default --yes || true
##[debug]
##[debug]yarn swa deploy .output\public --env default
##[debug]'
##[debug]Loading env
Run # Delete old deployment (otherwise we get a spawn EBUSY error when running the workflow again)
##[debug]C:\Program Files\PowerShell\7\pwsh.EXE -command ". 'D:\a\_temp\962b834b-afe0-484e-8e84-9c35b7e02f5a.ps1'"
ERROR: Operation returned an invalid status 'Bad Request'

Welcome to Azure Static Web Apps CLI (1.1.2)

Deploying front-end files from folder:
  D:\a\JabRefOnline\JabRefOnline\.output\public

Consider providing api-language and version using --api-language and --api-version flags,
    otherwise default values apiLanguage: node and apiVersion: 16 will apply

Deploying to environment: default


Found configuration file:
  D:\a\JabRefOnline\JabRefOnline\staticwebapp.config.json

Deploying project to Azure Static Web Apps...
Could not find StaticSitesClient local binary
- Downloading https://swalocaldeploy.azureedge.net/downloads/1.0.023421/windows/StaticSitesClient.exe@1.0.023421
[swa] ✔ Downloading https://swalocaldeploy.azureedge.net/downloads/1.0.023421/windows/StaticSitesClient.exe@1.0.023421
- Preparing deployment. Please wait...

✖ 
✖ Deployment Failed :(
✖ Deployment Failure Reason: spawn EBUSY
✖ For further information, please visit the Azure Static Web Apps documentation at https://docs.microsoft.com/azure/static-web-apps/
✖ If you believe this behavior is unexpected, please raise a GitHub issue at:
  https://github.com/Azure/static-web-apps-cli/issues/new/choose
Error: Process completed with exit code 1.
##[debug]Finishing: Deploy Web App
```
e.g. https://github.com/JabRef/JabRefOnline/actions/runs/5480147768/jobs/9989069413

**To Reproduce**
Steps to reproduce the behavior:

1. Run `swa deploy` in CI.


**Expected behavior**
Deployment is successful. 

**Screenshots**
If applicable, add screenshots to help explain your problem.


**Desktop (please complete the following information):**

- OS: Github CI, Ubuntu
- Version [e.g. 22]

--- a/src/core/download-binary-helper.ts
+++ b/src/core/download-binary-helper.ts
@@ -1,6 +1,7 @@
 import chalk from "chalk";
 import crypto from "node:crypto";
 import fs from "node:fs";
+import stp from "node:stream/promises";
 import { fetchWithProxy as fetch } from "./utils/fetch-proxy.js";
 import ora from "ora";
 import path from "node:path";
@@ -41,44 +42,35 @@ export async function downloadAndValidateBinary(
 
   createBinaryDirectoryIfNotExists(id, outputFolder);
 
-  return await new Promise<string>((resolve, reject) => {
-    const isPosix = platform === "linux-x64" || platform === "osx-x64";
-    const outputFile = path.join(outputFolder, id, downloadFilename);
+  const isPosix = platform === "linux-x64" || platform === "osx-x64";
+  const outputFile = path.join(outputFolder, id, downloadFilename);
 
-    const writableStream = fs.createWriteStream(outputFile, { mode: isPosix ? 0o755 : undefined });
-    bodyStream?.pipe(writableStream);
+  const writableStream = fs.createWriteStream(outputFile, { mode: isPosix ? 0o755 : undefined });
 
-    writableStream.on("end", () => {
-      bodyStream?.end();
-    });
+  await stp.pipeline(bodyStream, writableStream);
 
-    writableStream.on("finish", async () => {
-      const computedHash = computeChecksumfromFile(outputFile).toLowerCase();
-      const releaseChecksum = releaseMetadata.files[platform].sha.toLowerCase();
-      if (computedHash !== releaseChecksum) {
-        try {
-          // in case of a failure, we remove the file
-          fs.unlinkSync(outputFile);
-        } catch {
-          logger.silly(`Not able to delete ${downloadFilename}, please delete manually.`);
-        }
+  const computedHash = computeChecksumfromFile(outputFile).toLowerCase();
+  const releaseChecksum = releaseMetadata.files[platform].sha.toLowerCase();
+  if (computedHash !== releaseChecksum) {
+    try {
+      // in case of a failure, we remove the file
+      fs.unlinkSync(outputFile);
+    } catch {
+      logger.silly(`Not able to delete ${downloadFilename}, please delete manually.`);
+    }
 
-        spinner.fail();
-        reject(new Error(`Checksum mismatch for ${binaryName}! Expected ${releaseChecksum}, got ${computedHash}.`));
-      } else {
-        spinner.succeed();
+    spinner.fail();
+    throw new Error(`Checksum mismatch for ${binaryName}! Expected ${releaseChecksum}, got ${computedHash}.`);
+  } else {
+    spinner.succeed();
 
-        logger.silly(`Checksum match: ${computedHash}`);
-        logger.silly(`Saved binary to ${outputFile}`);
+    logger.silly(`Checksum match: ${computedHash}`);
+    logger.silly(`Saved binary to ${outputFile}`);
 
-        saveMetadata(releaseMetadata, outputFile, computedHash, binaryName);
+    saveMetadata(releaseMetadata, outputFile, computedHash, binaryName);
+  }
 
-        resolve(outputFile);
-      }
-    });
-
-    // writableStream.close();
-  });
+  return outputFile;
 }
 
 /**

--- a/src/core/download-binary-helper.ts
+++ b/src/core/download-binary-helper.ts
@@ -5,7 +5,6 @@ import stp from "node:stream/promises";
 import { fetchWithProxy as fetch } from "./utils/fetch-proxy.js";
 import ora from "ora";
 import path from "node:path";
-import { PassThrough } from "node:stream";
 import { DATA_API_BUILDER_BINARY_NAME, DATA_API_BUILDER_FOLDER, DEPLOY_BINARY_NAME, DEPLOY_FOLDER } from "./constants.js";
 import { logger } from "./utils/logger.js";
 
@@ -38,8 +37,6 @@ export async function downloadAndValidateBinary(
     throw new Error(`Failed to download ${binaryName} binary from url ${url}. File not found (${response.status})`);
   }
 
-  const bodyStream = response?.body?.pipe(new PassThrough());
-
   createBinaryDirectoryIfNotExists(id, outputFolder);
 
   const isPosix = platform === "linux-x64" || platform === "osx-x64";
@@ -47,7 +44,7 @@ export async function downloadAndValidateBinary(
 
   const writableStream = fs.createWriteStream(outputFile, { mode: isPosix ? 0o755 : undefined });
 
-  await stp.pipeline(bodyStream, writableStream);
+  await stp.pipeline(response.body, writableStream);
 
   const computedHash = computeChecksumfromFile(outputFile).toLowerCase();
   const releaseChecksum = releaseMetadata.files[platform].sha.toLowerCase();

